Bottom: 82701973de20fd9ef1198cbcf8b704c0b18cce8e
Top:    90d25b8276c78150b68fd32e082334d1a97d451b
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-12-16 15:50:14 +0800

Refresh of macvtap-xdp-fast-rx-support

---

diff --git a/drivers/net/macvtap.c b/drivers/net/macvtap.c
index 5b4654a..280c9a0 100644
--- a/drivers/net/macvtap.c
+++ b/drivers/net/macvtap.c
@@ -45,7 +45,7 @@ struct macvtap_queue {
 	bool enabled;
 	struct list_head next;
 	struct skb_array skb_array;
-	struct xdp_array xdp_array;
+	struct ptr_ring xdp_array;
 };
 
 #define MACVTAP_FEATURES (IFF_VNET_HDR | IFF_MULTI_QUEUE)
@@ -537,6 +537,11 @@ static void macvtap_sock_destruct(struct sock *sk)
 	skb_array_cleanup(&q->skb_array);
 }
 
+static void macvtap_xdp_free(void *ptr)
+{
+	kfree(ptr);
+}
+
 static int macvtap_open(struct inode *inode, struct file *file)
 {
 	struct net *net = current->nsproxy->net_ns;
@@ -581,7 +586,7 @@ static int macvtap_open(struct inode *inode, struct file *file)
 	if (skb_array_init(&q->skb_array, dev->tx_queue_len, GFP_KERNEL))
 		goto err_array;
 
-	if (ptr_ring_init(&q->xdp_array, dev->xdp_array, GFP_KERNEL))
+	if (ptr_ring_init(&q->xdp_array, dev->tx_queue_len, GFP_KERNEL))
 		goto err_xdp;
 
 	err = macvtap_set_queue(dev, file, q);
@@ -594,7 +599,7 @@ static int macvtap_open(struct inode *inode, struct file *file)
 	return err;
 
 err_queue:
-	ptr_ring_cleanup(&q->xdp_array, kfree);
+	ptr_ring_cleanup(&q->xdp_array, macvtap_xdp_free);
 err_xdp:
 	skb_array_cleanup(&q->skb_array);
 err_array:
